### PR TITLE
Warn when using a static class in an interface param or return

### PR DESCRIPTION
**Pull request by 333fred**
The native compiler had a bug here that permitted static classes to be used in parameter or return types for methods defined in an interface. We will now warn about doing this when warning waves are on.

Fixes https://github.com/dotnet/roslyn/issues/38256.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -2093,9 +2093,15 @@ If such a class is used as a base class and if the deriving class defines a dest
   <data name="ERR_ParameterIsStaticClass" xml:space="preserve">
     <value>'{0}': static types cannot be used as parameters</value>
   </data>
+  <data name="WRN_ParameterIsStaticClass" xml:space="preserve">
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:2096]: You'll probably need a corresponding `_Title` entry

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs [hunk 2]
```diff
@@ -254,7 +254,7 @@ private UnboundLambda BindAnonymousFunction(AnonymousFunctionExpressionSyntax sy
                     var type = lambda.Data.ParameterTypeWithAnnotations(i);
                     if (type.HasType && type.IsStatic)
                     {
-                        Error(diagnostics, ErrorCode.ERR_ParameterIsStaticClass, syntax, type.Type);
+                        Error(diagnostics, ErrorFacts.GetStaticClassParameterCode(isInInterface: false), syntax, type.Type);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:257]: nit: I'd probably would have kept using the error code directly, but that's not crucial...
Did you check we have existing tests for all the scenarios that are now using the two helper methods?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:257]: I did check for the other scenarios. I wanted to use the new helper in all locations in order to make it obvious to future readers that we'd looked at everywhere that reported the warning and made an explicit decision, not that it was forgotten when implemented.

---
In reply to: [474303458](https://github.com/dotnet/roslyn/pull/47012#discussion_r474303458) [](ancestors = 474303458)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs [hunk 3]
```diff
@@ -547,5 +549,11 @@ internal static bool PreventsSuccessfulDelegateConversion(ImmutableArray<Diagnos
 
             return false;
         }
+
+        internal static ErrorCode GetStaticClassParameterCode(bool isInInterface)
+            => isInInterface ? ErrorCode.WRN_ParameterIsStaticClass : ErrorCode.ERR_ParameterIsStaticClass;
+
+        internal static ErrorCode GetStaticClassReturnCode(bool useWarning)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:556]: nit: why do these parameters have different names..?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:556]: Because I forgot to rename the other one :)

**Comment by 333fred** [comment]
@RikkiGibson do you have any more comments?

**Comment by 333fred** [comment]
@RikkiGibson do you have any more feedback?

**Comment by RikkiGibson** [comment]
Ah, sorry I missed the pings 😄 will take another look

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs [hunk 4]
```diff
@@ -13334,6 +13334,39 @@ public class Test
...
+        [Fact, WorkItem(38256, "https://github.com/dotnet/roslyn/issues/38256")]
+        public void ParameterAndReturnTypesAreStaticClassesWarning()
+        {
+            var source = @"
+static class C {}
+interface I
+{
+    void M1(C c); // 1
+    C M2(); // 2
+    C Prop { get; set; } // 3, 4
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs:13346]: Consider testing an indexer with a static class parameter

**Pull request merged**